sol/main.c

```diff
@@ -0,0 +1,507 @@
+		req_queue->highest_priority = 0;
+		req_queue->lowest_priority = GK_NUM_REQ_PRIORITIES;
+	} else {
+		struct priority_req *first = list_first_entry(&req_queue->head,
+			struct priority_req, list);
+		req_queue->highest_priority = first->priority;
+	}
+
+	/* We cannot drop the packets, so re-send. */
+	do {
+		uint16_t sent = rte_eth_tx_burst(tx_port, tx_queue,
+			pkts_out + total_sent, nb_pkts_out);
+		total_sent += sent;
+		nb_pkts_out -= sent;
+	} while (nb_pkts_out > 0);
+}
+
+static int
+req_queue_init(struct sol_config *sol_conf)
+{
+	struct req_queue *req_queue = &sol_conf->req_queue;
+	struct rte_eth_link link;
+	uint32_t cycles_per_byte;
+	uint64_t rate;
+	int ret;
+
+	req_queue->len = 0;
+	req_queue->highest_priority = 0;
+	req_queue->lowest_priority = GK_NUM_REQ_PRIORITIES;
+
+	/* Token bucket. */
+	ret = rte_approx(REQ_QUEUE_BW_RATE, GATEKEEPER_TB_RATE_CONFIG_ERR,
+		&req_queue->tb_credits_per_period, &req_queue->tb_period_bytes);
+	if (ret < 0) {
+		RTE_LOG(ERR, GATEKEEPER, "sol: could not approximate the request queue's allocated bandwidth\n");
+		return ret;
+	}
+
+	/* Calculate link speed in bytes per second. */
+	rte_eth_link_get(sol_conf->net->back.id, &link);
+	if (link.link_speed == ETH_SPEED_NUM_NONE) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"sol: front interface link speed is undefined\n");
+		return -1;
+	}
+	rate = (uint64_t)link.link_speed * 1000 * 1000 / 8;
+	cycles_per_byte = (rte_get_tsc_hz() << GATEKEEPER_TIME_SHIFT) / rate;
+	req_queue->inv_cycles_per_byte = rte_reciprocal_value(cycles_per_byte);
+
+	/* The DPDK sched library starts the bucket half full. */
+	req_queue->tb_size = rate * REQ_QUEUE_BW_RATE;
+	req_queue->tb_credits = req_queue->tb_size / 2;
+	req_queue->tb_time_bytes = 0;
+
+	req_queue->time_cpu_cycles = rte_get_tsc_cycles();
+	req_queue->time_cpu_bytes = 0;
+	req_queue->time_bytes = 0;
+
+	INIT_LIST_HEAD(&req_queue->head);
+	memset(req_queue->priorities, 0, sizeof(req_queue->priorities));
+
+	return 0;
+}
+
+static int
+cleanup_sol(struct sol_config *sol_conf)
+{
+	struct req_queue *req_queue = &sol_conf->req_queue;
+	struct priority_req *entry, *next;
+
+	list_for_each_entry_safe(entry, next, &req_queue->head, list) {
+		rte_pktmbuf_free(entry->pkt);
+		list_del(&entry->list);
+		mb_free_entry(&sol_conf->mb, entry);
+		req_queue->len--;
+	}
+
+	if (req_queue->len > 0)
+		RTE_LOG(NOTICE, GATEKEEPER, "sol: bug: removing all requests from the priority queue on cleanup leaves the queue length at %"PRIu32"\n",
+			req_queue->len);
+
+	destroy_mailbox(&sol_conf->mb);
+	rte_free(sol_conf);
+	return 0;
+}
+
+static int
+sol_proc(void *arg)
+{
+	struct sol_config *sol_conf = (struct sol_config *)arg;
+	unsigned int lcore = sol_conf->lcore_id;
+	uint8_t tx_port_back = sol_conf->net->back.id;
+
+	RTE_LOG(NOTICE, GATEKEEPER,
+		"sol: the Solicitor block is running at lcore = %u\n", lcore);
+
+	while (likely(!exiting)) {
+		enqueue_reqs(sol_conf);
+		dequeue_reqs(sol_conf, tx_port_back,
+			sol_conf->tx_queue_back);
```

Comment:
Drop the third parameter of `dequeue_reqs()`; `dequeue_reqs()` can read `sol_conf->tx_queue_back` thanks to its first parameter. There's no gain in having the third parameter.

The second parameter is okay.